Automattic/newspack-extended-access
# ✨ Add support for Single Post Subscription

The goal of the PR is to add support for Single Post Subscription by utilizing Google Extended Access and COOKIE based authentication for the post.

![Google Extended Access Prompt](https://user-images.githubusercontent.com/75766877/235243163-9513b5a7-0394-4d28-8f5c-a8d1b6e31318.png)

### Changes proposed in this Pull Request:

* [x] 📜 Register Google Extended Script and Callbacks.
* [x] ⚙️ Add Option Tab (`Dashboard -> WooCommerce -> Settings -> Memberships -> Newspack Extended Access`).
* [x] ✅ Checks to verify valid dependencies.
* [x] ⛔️ Show admin notice when dependencies do not met or configured incorrectly.
* [x] 🧩 Google Prompts
  * [x] 🔐 Display **Google Login Prompt** when no user is logged.
  * [x] 🔏 Display **Google Subscribe Prompt** when user is logged in but not subscribed to the post.
* [x] 🍪 `COOKIE` based authentication (based on `'newspack_' . md5( $post_id . $user_id )`).
  * [x] Google Extended Access Script Registration
  * [x] REST Controller
  * [ ] Dependency Manager (tests cannot be performed due to dependency plugin `WooCommerce Memberships`)
  * [ ] Single Post Subscription controller (tests cannot be performed due to dependency plugin `WooCommerce Memberships`)

## How to test the changes in this Pull Request

1. Change branch to this PR branch.
2. Install dependencies `composer install`
3. Prepare zip and upload to site OR directly copy content of the repository to you site.
4. Make sure to [install and configure dependencies](#dependency).
5. Configure [Google Client API ID](#configuration).
6. Goto one of the premium post and make sure to add test URL parameters:
  `?gaa_at=testp&gaa_n=1234567&gaa_sig=testsig&gaa_ts=999999999`

> **Important Note**
> Account with `Administrator` or `Editor` role cannot subscribe. This is [Newspack restriction](https://github.com/Automattic/newspack-plugin/blob/a1d4eb7aad13449cb1d847fb25ee2414028fdb69/includes/reader-activation/class-reader-activation.php#L628-L636) thus users cannot login via Extended Access. Hence, you need another Google account to login & subscribe to a post via Extended Access dialog.

## UX Flow

### Scene 1: Anonymous User (New User)

1. User opens a premium post (that includes `?gaa_at=testp&gaa_n=1234567&gaa_sig=testsig&gaa_ts=999999999` parameters embedded by Google News).
2. Google Extended Access requests REST Controller to get logged in user.
  a. REST Controller responds with `NO_USER_LOGGED_IN`.
  b. Google Extended Access now displays `Google Login Prompt`.
  c. User login through `Google Login Prompt` (and user gets automatically registered by controller as `customer`).
3. Google Extended Access now ask REST Controller for user's subscription status for the post (by verifying COOKIE).
  a. REST Controller responds with `NO_SUBSCRIPTION` found for user for the post.
  b. Google Extended Access displays `Google Subscription Prompt`.
  c. Here, if user click on `Subscribe` button, a COOKIE on the machine will be created and Google Extended Access unlocks article.
  d. If user click `X` button, `Google Subscription Prompt` went away.

### Scene 2: Logged in User with No Subscription

1. User opens a premium post (that includes `?gaa_at=testp&gaa_n=1234567&gaa_sig=testsig&gaa_ts=999999999` parameters embedded by Google News).
2. Google Extended Access requests REST Controller to get logged in user.
  a. REST Controller responds with `LOGGED_IN` user but post `NO_SUBSCRIPTION`.
3. Google Extended Access now ask REST Controller for user's subscription status for the post (by verifying COOKIE).
  a. REST Controller responds with `NO_SUBSCRIPTION` found for user for the post.
  b. Google Extended Access displays `Google Subscription Prompt`.
  c. Here, if user click on `Subscribe` button, a COOKIE on the machine will be created and Google Extended Access unlocks article.
  d. If user click `X` button, `Google Subscription Prompt` went away.

### Scene 3: Registered in User with Subscription

1. User opens a premium post (that includes `?gaa_at=testp&gaa_n=1234567&gaa_sig=testsig&gaa_ts=999999999` parameters embedded by Google News).
2. Google Extended Access requests REST Controller to get logged in user.
  a. REST Controller responds with `NO_USER_LOGGED_IN`.
  b. Google Extended Access now displays `Google Login Prompt`.
  c. User login through `Google Login Prompt`.
3. Google Extended Access now ask REST Controller for user's subscription status for the post (by verifying COOKIE).
  a. REST Controller responds with `VALID_SUBSCRIPTION` found for user for the post.
  b. Google Extended Access unlocks article.

### Scene 4: Logged in User with Subscription

1. User opens a premium post (that includes `?gaa_at=testp&gaa_n=1234567&gaa_sig=testsig&gaa_ts=999999999` parameters embedded by Google News).
2. Google Extended Access requests REST Controller to get `LOGGED_IN` user.
  a. REST Controller responds with logged in user with `VALID_SUBSCRIPTION`.
  b. Google Extended Access unlocks article.

## Dependency

This plugin requires following plugin to be installed, configured and active:
- [Newspack Plugin](https://github.com/Automattic/newspack-plugin)
- [WooCommerce](https://github.com/woocommerce/woocommerce)
- [WooCommerce Memberships](https://woocommerce.com/products/woocommerce-memberships/)

## Configuration

This plugin requires a valid `Google Client API ID` specific to your Newspack Site domain. Please create a new `Google Client API ID` for your site, if not already created. Follow instruction provided on [Setting up Google Client API ID]([https://](https://developers.google.com/identity/gsi/web/guides/get-google-api-clientid)).

> Make sure [dependent plugins](#dependency) are installed, configured and active on your site.

### Setting up Google Client API ID

Login to you Newspack site dashboard as an `Administrator`:
1. Open `WooCommerce` menu
9. Select `Settings` menu
10. Select `Memberships` tab
11. Select sub-tab `Newspack Extended Access`
12. Add your valid `Google Client API ID`

![Configure Google Client API ID](https://user-images.githubusercontent.com/75766877/235243611-c654653a-be35-4437-80c8-c09289bbf8ff.png)
